source/simulator/core.cc

```diff
@@ -2421,6 +2399,11 @@ namespace aspect
         // then do the core work: assemble systems and solve
         solve_timestep ();
 
+        const bool statistics_already_generated = (timestep_number == 0) && (pre_refinement_step == 0)
+            && (parameters.nonlinear_solver == NonlinearSolver::IMPES || parameters.nonlinear_solver == NonlinearSolver::Advection_only);
+        if (!statistics_already_generated)
+          generate_global_statistics();
```

Comment:
This is complicated to follow. If you are in time step zero, you say that the statistics have already been generated. I suspect that is because further down, you actually generate them in the `initialize_` function, but that seems like a difficult to follow step for a reader. I'll ask more questions in the general comments.